### ISSUE TITLE
YOLO-v8-POSE visibility feature in prediction output

### ISSUE THREAD

**Issue by jrajpal5-singularity** [ultralytics/ultralytics#8242]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

The YOLO-V8-POSE model only predicts keypoints even if visibility is mentioned in input dataset which makes the visibility information useless as the time of inference.

By addition of Visibility feature; the model would have to predict one additional value with each key point; but this hand-off would help in multiple ways.
1) Better tracking of each keypoint which may help in many applications.
2) Auto-annotation pipeline would be improved.  

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@jrajpal5-singularity hello! Thank you for bringing up the visibility feature for the YOLOv8-POSE model. We appreciate your suggestion and understand how predicting visibility alongside keypoints could enhance the model's utility for certain applications.

Currently, the YOLOv8-POSE model is designed to predict keypoints without visibility information. However, we recognize the potential benefits of including visibility predictions. While we don't have this feature at the moment, we'll consider it for future updates.

In the meantime, if you're looking to implement this yourself, you might consider modifying the model head to output an additional value for visibility and adjust the loss function accordingly. Here's a very basic example of how you might approach this in the model definition:

```python
# Assuming 'model' is your YOLOv8-POSE model instance
# and 'visibility' is a tensor representing visibility scores

class CustomPoseHead(nn.Module):
    def __init__(self, original_head):
        super().__init__()
        self.original_head = original_head
        # Add an additional layer for visibility prediction
        self.visibility_layer = nn.Conv2d(in_channels=..., out_channels=1, kernel_size=...)

    def forward(self, x):
        keypoints = self.original_head(x)
        visibility = self.visibility_layer(x)
        return torch.cat((keypoints, visibility), dim=1)

model.head = CustomPoseHead(model.head)
```

Please note that this is a simplified example and actual implementation would require a more in-depth modification of the model architecture and training pipeline.

We're always looking to improve and your feedback is valuable. If you have any further insights or questions, feel free to share! 😊👍

**Comment by glenn-jocher** [comment]
@jrajpal5-singularity hi there! 👋 Thank you for bringing up the visibility feature for keypoints in YOLOv8-POSE. We appreciate your suggestion and understand how predicting visibility alongside keypoints can be beneficial for various applications, including tracking and auto-annotation.

Currently, YOLOv8-POSE does not output visibility information, but we recognize the potential improvements it could bring. We'll consider this feature for future updates. In the meantime, if you're comfortable with modifying the model, you could potentially implement this by adjusting the output layer to include visibility predictions and modifying the loss function accordingly.

For example, you might adjust the model's head to output an additional channel for visibility scores and then apply a suitable loss function that accounts for the visibility during training.

```python
# Pseudo code for adding visibility to the model's output
class CustomPoseHead(nn.Module):
    def __init__(self, ...):
        # initialize layers
        self.visibility_output = nn.Conv2d(...)

    def forward(self, x):
        # your forward pass
        visibility_scores = self.visibility_output(x)
        return keypoints, visibility_scores
```

We'd love to see community contributions, so if you're interested in working on this feature, please feel free to discuss it further with us or submit a PR. We're here to support you through the process. 😊

Thanks again for your input! 🌟
